src/main/java/uk/gov/hmcts/probate/services/persistence/services/SequenceNumberService.java

```diff
@@ -1,25 +1,94 @@
 
 @Service
+@Transactional
 public class SequenceNumberService {
 
-    @Autowired
     private Map<String, PostgreSQLSequenceMaxValueIncrementer> registrySequenceNumbers;
+    private RegistryRepository registryRepository;
+    private EntityManager entityManager;
+    private final ObjectMapper mapper;
+
+    private static final String SUBMISSION_REFERENCE = "submissionReference";
+    private static final String REGISTRY = "registry";
+    private static final String SEQUENCE_NUMBER = "sequenceNumber";
+    private static final String NAME = "name";
+    private static final String ADDRESS = "address";
+    private static final String EMAIL = "email";
 
-    public Long getNext(String registryName) {
+    @Autowired
+    public SequenceNumberService(Map<String, PostgreSQLSequenceMaxValueIncrementer> registrySequenceNumbers,
+                                 RegistryRepository registryRepository, EntityManager entityManager,
+                                 ObjectMapper mapper) {
+        this.registrySequenceNumbers = registrySequenceNumbers;
+        this.registryRepository = registryRepository;
+        this.entityManager = entityManager;
+        this.mapper = mapper;
+    }
+
+    public JsonNode getNextRegistry(long submissionReference) {
+        Registry nextRegistry = identifyNextRegistry();
+        nextRegistry.incrementCounter();
+        entityManager.lock(registryRepository.save(nextRegistry), LockModeType.PESSIMISTIC_WRITE);
+        return populateRegistrySubmitData(submissionReference, nextRegistry);
+    }
 
+    public Long getNextSequenceNumber(String registryName) {
         try {
             return registrySequenceNumbers.get(registryName).nextLongValue();
         } catch (NullPointerException e) {
-                throw new RegistryNotConfiguredException("Registry not configured: " + registryName, e);
+            throw new RegistryNotConfiguredException("Registry not configured: " + registryName, e);
         }
     }
+
+    JsonNode populateRegistrySubmitData(long submissionReference, Registry registry) {
+        ObjectNode registryDataObject = mapper.createObjectNode();
+        ObjectNode registryMapper = mapper.createObjectNode();
+
+        registryDataObject.put(SUBMISSION_REFERENCE, submissionReference);
+        registryMapper.put(NAME, registry.capitalizeRegistryName());
+        registryMapper.put(SEQUENCE_NUMBER, getNextSequenceNumber(registry.getId()));
+        registryMapper.put(EMAIL, registry.getEmail());
+        registryMapper.put(ADDRESS, registry.getAddress());
+        registryDataObject.set(REGISTRY, registryMapper);
+
+        return registryDataObject;
+    }
+
+    Registry identifyNextRegistry() {
+        List<Registry> registries = registryRepository.findAll().stream()
+                .filter(r -> r.getCounter() < r.getRatio())
+                .collect(Collectors.toList());
+        return registries.isEmpty() ?
+                resetRegistryCounters().get(0) : registries.get(0);
```

Comment:
Only observation here is that the registries will be picked in one go until the ratio is reached and then the next registry will be picked until the list it empty and the counters reset. It might be nicer to spread the registry selection by changing 'registries.get(0);' to select any of the registries listed as random?

Reply:
I see where you are coming from and would be nicer. However the amount of code required to implement would bloat that function.